### PR TITLE
(feat): add comprehensive tests suite (phase 1)

### PR DESCRIPTION
**Pull request by kosa12**
## Summary by CodeRabbit

* **New Features**
  * Added support for custom server type prefix alongside existing server formats.

* **Tests**
  * Expanded comprehensive test coverage across all major components and workflows to improve reliability and catch regressions earlier.

* **Chores**
  * Updated development environment tooling and CI/build workflows for better test infrastructure and faster feedback cycles.

[Review Change Stack](https://app.coderabbit.ai/change-stack/Kopuz-org/kopuz/pull/269?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Comment by temidaradev** [comment]
what type of warmup is this

**Comment by kosa12** [comment]
> what type of warmup is this

Since the test suite and the 2 builds run on nix, there is no need to download all the stuff 3 times, download and upackage once, upload to cache, fan out to the 3 other jobs

image

**Comment by temidaradev** [comment]
mmm i see

**Comment by kosa12** [comment]
image

**Inshallah, there won't be regression after this**

**Comment by temidaradev** [comment]
hell yeah

File: crates/reader/src/utils.rs [hunk 1]
```diff
@@ -64,3 +64,97 @@ pub fn save_cover(
...
+
+    #[test]
+    fn detect_image_extension_falls_back_to_jpg_for_unknown_bytes() {
+        assert_eq!(detect_image_extension(b"not-an-image"), "jpg");
+        assert_eq!(detect_image_extension(b""), "jpg");
+        assert_eq!(detect_image_extension(b"12"), "jpg"); // Very short
+    }
+
+    #[test]
+    fn find_folder_cover_prefers_cover_before_folder_variants() {
```
Comment (Irilith) [inline crates/reader/src/utils.rs]: can we have more test for this which contain some edge case (no extension and more as you can see)? 
something like 
```rust
    fn setup() -> TempDir {
        tempfile::tempdir().unwrap()
    }

    fn touch(dir: &std::path::Path, name: &str) {
        File::create(dir.join(name)).unwrap();
    }

    #[test]
    fn finds_cover_jpg() {
        let tmp = setup();
        touch(tmp.path(), "cover.jpg");
        assert!(find_folder_cover(tmp.path()).is_some());
    }

    #[test]
    fn finds_case_insensitive_stem() {
        let tmp = setup();
        touch(tmp.path(), "Cover.png");
        assert!(find_folder_cover(tmp.path()).is_some());
    }

    #[test]
    fn finds_case_insensitive_ext() {
        let tmp = setup();
        touch(tmp.path(), "cover.JPG");
        assert!(find_folder_cover(tmp.path()).is_some());
    }

    #[test]
    fn finds_folder_and_album_stems() {
        let tmp = setup();
        touch(tmp.path(), "folder.png");
        assert!(find_folder_cover(tmp.path()).is_some());

        let tmp2 = setup();
        touch(tmp2.path(), "album.webp");
        assert!(find_folder_cover(tmp2.path()).is_some());
    }

    #[test]
    fn no_match_wrong_stem() {
        let tmp = setup();
        touch(tmp.path(), "artwork.jpg");
        assert!(find_folder_cover(tmp.path()).is_none());
    }

    #[test]
    fn no_match_wrong_ext() {
        let tmp = setup();
        touch(tmp.path(), "cover.bmp");
        assert!(find_folder_cover(tmp.path()).is_none());
    }

    #[test]
    fn no_extension_skipped() {
        let tmp = setup();
        touch(tmp.path(), "cover"); // no extension
        assert!(find_folder_cover(tmp.path()).is_none());
    }

    #[test]
    fn extra_dot_in_stem_no_match() {
        let tmp = setup();
        touch(tmp.path(), "cover.old.jpg"); // stem = "cover.old"
        assert!(find_folder_cover(tmp.path()).is_none());
    }

    #[test]
    fn empty_dir_returns_none() {
        let tmp = setup();
        assert!(find_folder_cover(tmp.path()).is_none());
    }

    #[test]
    fn nonexistent_dir_returns_none() {
        let path = std::path::Path::new("/tmp/does_not_exist_kopuz_test");
        assert!(find_folder_cover(path).is_none());
    }

    #[test]
    fn subdirectory_named_cover_jpg_ignored() {
        let tmp = setup();
        fs::create_dir(tmp.path().join("cover.jpg")).unwrap(); // dir, not file
        assert!(find_folder_cover(tmp.path()).is_none());
    }

    #[test]
    fn returns_correct_path() {
        let tmp = setup();
        touch(tmp.path(), "folder.jpeg");
        let result = find_folder_cover(tmp.path()).unwrap();
        assert_eq!(result.file_name().unwrap(), "folder.jpeg");
    }

    #[test]
    fn symlink_to_valid_image_found() {
        let tmp = setup();
        let dir = tmp.path();

        let real = tmp.path().join("real_cover.jpg");
        File::create(&real).unwrap();

        // symlink cover.jpg -> real_cover.jpg
        std::os::unix::fs::symlink(&real, dir.join("cover.jpg")).unwrap();

        assert!(find_folder_cover(dir).is_some());
    }

    #[test]
    fn broken_symlink_skipped() {
        let tmp = setup();
        let dir = tmp.path();

        // symlink pointing to nothing
        std::os::unix::fs::symlink("/nonexistent/cover.jpg", dir.join("cover.jpg")).unwrap();

        assert!(find_folder_cover(dir).is_none());
    }
```

Comment (kosa12) [reply crates/reader/src/utils.rs]: Sure, that was my idea as well, just didn't had time to work on this yesterday

**Comment by temidaradev** [comment]
bozo hates github so much that he rapes the ci

**Comment by kosa12** [comment]
> bozo hates github so much that he rapes the ci

They have to work for that 0.97 uptime

**Comment by NotAShelf** [comment]
The test result script is unnecessary as `cargo test` reports results perfectly well. If we want a faster test runner and structured results, I suggest cargo-nextest instead of slopping something out.

**Comment by kosa12** [comment]
> The test result script is unnecessary as `cargo test` reports results perfectly well. If we want a faster test runner and structured results, I suggest cargo-nextest instead of slopping something out.

I know, was thinking about it, just didn't want to bring a whole new dependency into the project, i just want a clean one-line summary at the end

**Comment by kosa12** [comment]
> > The test result script is unnecessary as `cargo test` reports results perfectly well. If we want a faster test runner and structured results, I suggest cargo-nextest instead of slopping something out.
> 
> I know, was thinking about it, just didn't want to bring a whole new dependency into the project, i just want a clean one-line summary at the end

Eh, you know what, fuck it

**Comment by kosa12** [comment]
> The test result script is unnecessary as `cargo test` reports results perfectly well. If we want a faster test runner and structured results, I suggest cargo-nextest instead of slopping something out.

will integrate nextest in phase 2

**Comment by NotAShelf** [comment]
`nix-shell -p cargo-nextest` :trollface:

**Comment by temidaradev** [comment]
im leaving this to raf

**Comment by NotAShelf** [comment]
I'll take a look tomorrow morning.

**Review by NotAShelf — COMMENTED** [review]
Okay I'm not really sure how to approach this tbh. It looks like those tests test *reimplementations*, not production code. This is the biggest problem I see, and a classic AI slop pattern if I may add. 

There are inline "helper functions" inside test blocks that are stripped down **copies** of what the Dioxus components do at runtime. There are a few good tests but a majority are... bogus.

**Comment by kosa12** [comment]
> Okay I'm not really sure how to approach this tbh. It looks like those tests test _reimplementations_, not production code. This is the biggest problem I see, and a classic AI slop pattern if I may add.
> 
> There are inline "helper functions" inside test blocks that are stripped down **copies** of what the Dioxus components do at runtime. There are a few good tests but a majority are... bogus.

Yeag. Those were added because the logic in some of the files are scattered everywhere, and since this is just a test PR, i wouldn't want to refactor it. I'll remove them and will add them back as integration tests.

File: packaging/nix/shell.nix [hunk 2]
```diff
@@ -10,6 +10,7 @@
   appstream,
   nodejs_22,
   yt-dlp,
+  deno,
```
Comment (Irilith) [inline packaging/nix/shell.nix]: didnt we have problem with deno in #288 
https://github.com/Kopuz-org/kopuz/commit/2cd4a8f89f620fc0fcff4661ddcbc431cde34be4

Comment (NotAShelf) [reply packaging/nix/shell.nix]: I reckon this is a bad rebase. Deno is otherwise unnecessary.

Comment (kosa12) [reply packaging/nix/shell.nix]: Yepp, got added by a merge conflict

**Comment by kosa12** [comment]
can i merge this? or shall i revisit again?

**Comment by kosa12** [comment]
> can i merge this? or shall i revisit again?

I will merge this today if no remonstrance

**Comment by NotAShelf** [comment]
I'd rather you clean up the git history first and let us review *properly*. 3000 lines of code is not something you should expect reviewed in a day.

**Comment by kosa12** [comment]
> I'd rather you clean up the git history first and let us review _properly_. 3000 lines of code is not something you should expect reviewed in a day.

okay

**Review by temidaradev — CHANGES_REQUESTED** [review]
conflicts :(

**Comment by UMCEKO** [comment]
even more conflicts :/
